File: DailyChallenges/CPP/LeetCode_1971.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        if (source == destination) return true;
        unordered_map<int, vector<int>> m;
        for (auto e: edges)
            m[e[0]].push_back(e[1]), m[e[1]].push_back(e[0]);

        vector<bool> visited(n, false); visited[source] = true;
        queue<int> q; q.push(source);
        while (!q.empty()){
            int sz = q.size();
            for (int i=0; i<sz; ++i){
                int src = q.front(); q.pop();
                for (auto des: m[src]){
                    if (des == destination) return true;
                    if (!visited[des]){
                        q.push(des);
                        visited[des] = true;
                    }
                }
            }
        }
        return false;
        // function<bool(int)> dfs = [&](int src){
        //     visited[src] = true;
        //     if (src == destination) return true;
        //     for (auto& des: m[src])
        //         if (!visited[des] && dfs(des)) return true;
        //     return false;
        // };
        // return dfs(source);
    }
};

File: DailyChallenges/Scala/LeetCode_1971.scala
object Solution {
    def validPath(n: Int, edges: Array[Array[Int]], start: Int, end: Int): Boolean = {
        if (start == end) return true
        val g = collection.mutable.HashMap.empty[Int,List[Int]]
        for (a <- edges){
            g(a(0)) = a(1) :: g.getOrElse(a(0),Nil)
            g(a(1)) = a(0) :: g.getOrElse(a(1),Nil)
        }
        
        val reach = Array.fill[Boolean](n)(false); reach(start) = true 
        val q = collection.mutable.Queue(start)
        while (q.nonEmpty) {
            val u = q.dequeue
            for (v <- g.getOrElse(u,Nil) if !reach(v)){
                if (v == end) return true 
                q.enqueue(v)
                reach(v) = true 
            }
        }
        false
    }
}
